server/routes/visitors_all.js

```diff
@@ -1,10 +1,53 @@
 const router = require('express').Router();
+const { filter, identity, pipe, omit, assoc } = require('ramda');
+const Joi = require('joi');
+const { validate } = require('../shared/middleware');
 const visitorsAll = require('../database/queries/visitors_all');
+const { ageRange, renameKeys } = require('../shared/util/helpers');
 
-router.post('/', (req, res, next) => {
-  visitorsAll(req.app.get('client:psql'), req.auth.cb_id)
-    .then(users => res.send({ result: users }))
-    .catch(next);
+const schema = {
+  query: {
+    offset: Joi.number()
+      .integer()
+      .min(0)
+      .required(),
+    gender: Joi.any().valid('', 'male', 'female', 'prefer not to say'),
+    age: Joi.any().valid('', '0-17', '18-34', '35-50', '51-69', '70+'),
+    activity: Joi.string().allow(''),
+  },
+};
+
+const removeProperties = omit(['offset', 'age']);
+const removeEmpty = filter(identity);
+
+router.get('/', validate(schema), async (req, res, next) => {
+  try {
+    const { query } = req;
+    const addCbId = assoc('activities.cb_id', req.auth.cb_id);
+
+    const createWhereObj = pipe(
+      renameKeys({ activity: 'activities.name', gender: 'sex' }),
```

Comment:
Doesn't rely on the request, can be defined outside of the handler.

Reply:
I'm creating the pipe in the handler as line 32 `addCbId` relies on the request. I think it's more readable to create the pipe in one go rather than creating a pipe that gets piped into another pipe as we may go pipe mad...

Reply:
I'm not talking about the pipe.

Reply:
You're talking about line 29 right? Which is defined externally and called with one arg in the handler.

More importantly... I've broken git... Help would be greatly appreciated.

Reply:
It can be entirely defined externally, it has no dependencies on the request. Removing unnecessary operations from handlers makes them more performant.

Reply:
So you're suggesting setting the partially applied function to a variable outside of the handler?

Reply:
Yeah, just like you did with all the other functions in the pipe